internal/aeintegrate/aeintegrate.go

```diff
@@ -0,0 +1,350 @@
+func (p *App) Deploy() error {
+	// Don't deploy unless we're certain everything is ready for deployment
+	// (i.e. admin client is authenticated and authorized)
+	if err := p.validate(); err != nil {
+		return err
+	}
+	if _, err := p.Module(); err != nil {
+		return fmt.Errorf("could not get module id: %v", err)
+	}
+	if err := p.setupAdmin(); err != nil {
+		return fmt.Errorf("could not setup admin service: %v", err)
+	}
+
+	log.Printf("(%s) Deploying...", p.Name)
+
+	cmd, err := p.deployCmd()
+	if err != nil {
+		log.Printf("(%s) Could not get deploy command: %v", p.Name, err)
+		return err
+	}
+
+	out, err := cmd.CombinedOutput()
+	// TODO: add a flag for verbose output (e.g. when running with binary created with `go test -c`)
+	if err != nil {
+		log.Printf("(%s) Output from deploy:", p.Name)
+		os.Stderr.Write(out)
+		// Try to clean up resources.
+		p.Cleanup()
+	} else if err == nil {
+		p.deployed = true
+		log.Printf("(%s) Deploy successful.", p.Name)
+	}
+	return err
+}
+
+// appyaml returns the path of the config file.
+func (p *App) appyaml() string {
+	appyaml := p.AppYaml
+	if appyaml == "" {
+		appyaml = "app.yaml"
+	}
+	return appyaml
+}
+
+// envappyaml writes the temporary configuration file if it does not exist already,
+// then returns the path of the temporary config file.
+func (p *App) envappyaml() (string, error) {
+	if p.tempAppYaml != "" {
+		return p.tempAppYaml, nil
+	}
+
+	base := p.appyaml()
+	tmp := "aeintegrate." + base
+
+	if len(p.Env) == 0 {
+		err := os.Symlink(filepath.Join(p.Dir, base), filepath.Join(p.Dir, tmp))
+		if err != nil {
+			return "", err
+		}
+		p.tempAppYaml = tmp
+		return p.tempAppYaml, nil
+	}
+
+	baseContents, err := ioutil.ReadFile(filepath.Join(p.Dir, base))
+	if err != nil {
+		return "", err
+	}
+
+	var c yaml.MapSlice
+	if err := yaml.Unmarshal(baseContents, &c); err != nil {
+		return "", err
+	}
+
+	for _, e := range c {
+		k, ok := e.Key.(string)
+		if !ok || k != "env_variables" {
+			continue
+		}
+
+		vals, ok := e.Value.(yaml.MapSlice)
+		if !ok {
+			return "", fmt.Errorf("expected MapSlice for env_variables")
+		}
+
+		for _, kv := range vals {
+			k, ok := kv.Key.(string)
+			if !ok {
+				return "", fmt.Errorf("env_variables/%#v should be a string", kv.Key)
+			}
+			if val, ok := p.Env[k]; ok {
+				kv.Value = val
+			}
+		}
+	}
+
+	out, err := yaml.Marshal(c)
+	if err != nil {
+		return "", err
+	}
+
+	if err := ioutil.WriteFile(filepath.Join(p.Dir, tmp), out, 0755); err != nil {
+		return "", err
+	}
+
+	p.tempAppYaml = tmp
+	return p.tempAppYaml, nil
+}
+
+func (p *App) deployCmd() (*exec.Cmd, error) {
+	gcloudBin := os.Getenv("GCLOUD_BIN")
+	if gcloudBin == "" {
+		gcloudBin = "gcloud"
+	}
+	aedeploy := os.Getenv("AEDEPLOY_BIN")
+	if aedeploy == "" {
+		aedeploy = "aedeploy"
+	}
+
+	appyaml, err := p.envappyaml()
+	if err != nil {
+		return nil, err
+	}
+
+	// NOTE: if the "preview" and/or "app" components are not available, and this is
```

Comment:
I don't think so. Probably a bug that should be filed with the gcloud team.

Shouldn't be able to run multiple component updates simultaneously.